### PR TITLE
CATTY-470 Add Brick category overview

### PR DESCRIPTION
**Pull request by zwerg4**
- Created a new ViewController for the category overview which opens when adding a new brick.
- ~~When adding a brick after one already added a brick, the last category opens and not the category overview.~~
- Fixed UI Tests
- Changed the Category "Variable" to "Data"

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catty/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [X] Include the name of the Jira ticket in the PR’s title
- [X] Verify that the Jira ticket is in the status *Ready for Development*
- [X] Include a summary of the changes plus the relevant context
- [X] Choose the proper base branch (*develop*)
- [X] Confirm that the changes follow the project’s coding guidelines
- [X] Verify that the changes generate no compiler or linter warnings
- [X] Perform a self-review of the changes
- [X] Verify to commit no other files than the intentionally changed ones
- [X] Include reasonable and readable tests verifying the added or changed behavior
- [X] Confirm that new and existing unit tests pass locally
- [X] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [X] Stick to the project’s git workflow (rebase and squash your commits)
- [X] Verify that your changes do not have any conflicts with the base branch
- [ ] After the PR, verify that all CI checks have passed
- [ ] Post a message in the *#catty* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

File: src/Catty/Views/Custom/CollectionViewCells/Buttons/CategoriesOverviewCollectionViewCell.swift [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+
+import UIKit
+
+class CategoriesOverviewCollectionViewCell: UICollectionViewCell {
```
Comment (m-herold) [inline src/Catty/Views/Custom/CollectionViewCells/Buttons/CategoriesOverviewCollectionViewCell.swift:25]: Please rename to `BrickCategoryOverviewCollectionViewCell` to match the name of the parent collection view. Also please move to parent directory since this cell is not related to `UIButton` or create a new subdirectory (e.g., _BrickCategoryOverview_).

File: src/Catty/ViewController/Continue&New/MaintainObject/MaintainScript/ScriptCollectionViewController.h [hunk 2]
```diff
@@ -24,10 +24,14 @@
...
 
 @class SpriteObject;
+@class BrickCategoryOverviewController;
 
 @interface ScriptCollectionViewController : BaseCollectionViewController<BrickCategoryViewControllerDelegate>
 @property (nonatomic, strong) SpriteObject *object;
 
 - (void)removeBrickOrScript:(id<BrickProtocol>)scriptOrBrick atIndexPath:(NSIndexPath*)indexPath;
 
+- (void)showBrickPickerOfCategory:(BrickCategoryOverviewController*)bsvc selectedCategory:(BrickCategory*)selectedCategory;
```
Comment (m-herold) [inline src/Catty/ViewController/Continue&New/MaintainObject/MaintainScript/ScriptCollectionViewController.h:34]: Please remove deprecated method from header.

File: src/Catty/ViewController/Continue&New/MaintainObject/MaintainScript/ScriptCollectionViewController.h [hunk 3]
```diff
@@ -24,10 +24,14 @@
 #import "BaseCollectionViewController.h"
 
 @class SpriteObject;
+@class BrickCategoryOverviewController;
```
Comment (m-herold) [inline src/Catty/ViewController/Continue&New/MaintainObject/MaintainScript/ScriptCollectionViewController.h:27]: Please remove deprecated forward declaration.

File: src/Catty/ViewController/Continue&New/MaintainObject/MaintainScript/ScriptCollectionViewController.m [hunk 4]
```diff
@@ -156,25 +155,14 @@ - (void)viewDidDisappear:(BOOL)animated
...
-        bcvc.delegate = self;
-        
-        [bsvc setViewControllers:@[bcvc]
-                       direction:UIPageViewControllerNavigationDirectionForward
-                        animated:NO
-                      completion:NULL];
-        
-        UINavigationController *navController = [[UINavigationController alloc] initWithRootViewController:bsvc];
-        [self presentViewController:navController animated:YES completion:NULL];
+            BrickCategoryOverviewController *bcoc = [[BrickCategoryOverviewController alloc] initWithNibName:nil bundle:nil scvc:self];
```
Comment (m-herold) [inline src/Catty/ViewController/Continue&New/MaintainObject/MaintainScript/ScriptCollectionViewController.m:161]: Please revise indentation.

File: src/CattyUITests/Extensions/XCTestCaseExtension.swift [hunk 5]
```diff
@@ -120,13 +122,23 @@ extension XCTestCase {
     }
 
     func findBrickSection(_ name: String, in app: XCUIApplication) {
-        let maxTries = 8
+        NSLog("findBrickSection: name: \(name)")
```
Comment (m-herold) [inline src/CattyUITests/Extensions/XCTestCaseExtension.swift:125]: Please remove log message.

File: src/Catty/ViewController/Continue&New/MaintainObject/MaintainScript/BrickCategoryViewController.h [hunk 6]
```diff
@@ -41,6 +41,7 @@
 @property(nonatomic, weak) id<BrickCategoryViewControllerDelegate> delegate;
 @property(nonatomic, readonly) NSArray *bricks;
 @property(nonatomic, weak) SpriteObject *spriteObject;
+@property (nonatomic, strong) NSArray<BrickCategory *> *categorieBricks;
```
Comment (m-herold) [inline src/Catty/ViewController/Continue&New/MaintainObject/MaintainScript/BrickCategoryViewController.h:44]: As far as I can see, this attribute is unused, right?

File: src/Catty/ViewController/Continue&New/MaintainObject/MaintainScript/BrickCategoryOverviewController.swift [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+import UIKit
+
+@objc class BrickCategoryOverviewController: UIViewController, UICollectionViewDelegate, UICollectionViewDataSource {
+
+    var brickCategoryOverviewCollectionView: UICollectionView!
+    var categegoriesBricks = [BrickCategory]()
+    var scvc: ScriptCollectionViewController?
+
+    // MARK: Init
+    @objc override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
```
Comment (m-herold) [inline src/Catty/ViewController/Continue&New/MaintainObject/MaintainScript/BrickCategoryOverviewController.swift:32]: I think this initializer is unused.

File: src/Catty/ViewController/Continue&New/MaintainObject/MaintainScript/BrickCategoryOverviewController.swift [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        categegoriesBricks = CatrobatSetup.self.registeredBrickCategories()
+        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
+    }
+
+    @objc(initWithNibName:bundle:scvc:) init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?, scvcP: ScriptCollectionViewController) {
```
Comment (m-herold) [inline src/Catty/ViewController/Continue&New/MaintainObject/MaintainScript/BrickCategoryOverviewController.swift:37]: Since this class is always initialized with `nil` for the first two parameters, meaning that the `nibName` and `bundle` parameters are unused, I would recommend to make a new, required initializer which only takes the `scriptCollectionViewController` as a parameter, since that controller also required that attribute to be set. By doing so, you can change that attribute to non-optional (you can forward `nil` to the super initializer for `nibName` and `bundle`).

**Review by m-herold — CHANGES_REQUESTED** [review]

File: src/Catty/ViewController/Continue&New/MaintainObject/MaintainScript/ScriptCollectionViewController.m [hunk 9]
```diff
@@ -156,25 +155,14 @@ - (void)viewDidDisappear:(BOOL)animated
...
+    if ([sender isKindOfClass:[UIBarButtonItem class]])
+    {
 
-        BrickSelectionViewController *bsvc = [[BrickSelectionViewController alloc]
-                                              initWithTransitionStyle:UIPageViewControllerTransitionStyleScroll
-                                              navigationOrientation:UIPageViewControllerNavigationOrientationHorizontal
-                                              options:@{
-                                                        UIPageViewControllerOptionInterPageSpacingKey : @20.f
-                                                        }];
-        BrickCategory *selectedCategory = self.lastSelectedCategory == nil ? [bsvc.categories firstObject] : self.lastSelectedCategory;
```
Comment (m-herold) [inline src/Catty/ViewController/Continue&New/MaintainObject/MaintainScript/ScriptCollectionViewController.m:167]: Please remove the deprecated attribute `lastSelectedCategory`.

**Review by m-herold — CHANGES_REQUESTED** [review]

File: src/Catty/ViewController/Continue&New/MaintainObject/MaintainScript/BrickCategoryOverviewController.swift (hunk 7 above)
Comment (zwerg4) [reply src/Catty/ViewController/Continue&New/MaintainObject/MaintainScript/BrickCategoryOverviewController.swift:32]: It's needed because its a required initialiser and the data of the brick is loaded here.

Comment (m-herold) [reply src/Catty/ViewController/Continue&New/MaintainObject/MaintainScript/BrickCategoryOverviewController.swift:32]: Yes but why do we need it here, while we are not really using it? Without this initializer you can make the `scvc` property non-optional which would make a lot of sense.

**Review by m-herold — APPROVED** [review]
Thanks!
